src/react/utils.js

```diff
@@ -58,6 +60,95 @@ export function valueIsClassComponent(realm: Realm, value: Value) {
   return false;
 }
 
+// a nested object of a React Element should be hoisted where all its properties are known
+// at evaluation time to be "static" or "pure". This means function values in scope, strings, numbers
+// and deeply nested object/arrays that share the same heuristics
+function canHoistObject(realm: Realm, object: ObjectValue, hoistedScope?: Generator): boolean {
+  if (isReactElement(object)) {
+    return canHoistReactElement(realm, object);
+  }
+  for (let [propName] of object.properties) {
+    let prop = Get(realm, object, propName);
+    if (prop instanceof AbstractValue) {
+      return false;
+    } else if (prop instanceof FunctionValue && prop.parent !== hoistedScope) {
+      return false;
```

Comment:
I'm not sure about this `hoistedScope` heuristic. We use additionalFunctions to serialize an entire React app behind a branch (as opposed to a CS app for example). Then we would use the React optimization inside the additionalFunction serialization. So there are really *two* "main" scopes. I'm not sure it is safe to assume that there are only two scopes. I think this can end up screwing us.

I think we might be able to hoist even more though.

I think that we can hoist local functions too. It doesn't really have to do anything with the scope itself. It depends on what the function closes over. So what you could do instead is visit the environment bindings in the function and see if all of those can also be hoisted. I.e. they don't have abstract values. That way you don't have to use the `hoistedScope` heuristic.

E.g. this `<div />` should be hoistable even though it is referencing a local function.

```js
function Foo() {
  var foo = {hello:'world'};
  return <div onClick={() => console.log(foo.hello)} />;
}
```

If you implement this by scanning the function properties and closed over environment you can avoid the `hoistedScope` check.